### ISSUE TITLE
Issue when focusing most recent window on a space

### ISSUE THREAD

**Issue by n0ibe** [koekeishiya/yabai#544]
I have the following commands to create a new space and move there the focused window, and to focus a space:

    alt + cmd - n : yabai -m space --create && \
                    yabai -m window --space last && yabai -m space --focus last && yabai -m window --focus recent
    alt - 1 : yabai -m space --focus 1 && yabai -m window --focus recent
    alt - 2 : yabai -m space --focus 2 && yabai -m window --focus recent

I would expect `--focus recent` to focus the most recent window on a space, however sometimes that doesn't work. There are various situations where it doesn't work, but the simplest one that I've been able to reproduce is the following.

Start with 3 windows on space 1 (the only space for now), then issue the command bind to `alt + cmd - n` to create a second space, move the window there, focus the space and focus the window.

Now go back to space 1 with `alt - 1`, switch focus from one window to the other, then go back to space 2 with `alt - 2`. I find that the window is space 2 is now not focused, and if I go back to space 1 the focus switches between the two windows.

Am I using the wrong commands in the bindings or is this a bug? I'm on version 2.4.3.

**Comment by koekeishiya** [comment]
The recently focused window id is just a global storage, and is not stored per space, but the lookup is only performed in the active space. This means that if you are on space 1 and have window A and B opened, moving the focus from A to B, will make A the recent window. Focusing the recent window will try to lookup A by ID in the active space, and in this case it will find and focus it, and make B the recent window. If you now switch to space 2, A will become the recent window, and if you try to focus recent now it will fail, because A is not located in the currently active space.

So I agree that this a bug. Now the question is if we want *recent* to be space specific or global. My initial implementation was for this to be a global thing, but I think maybe the most recent *in* a space itself is more valuable, since we already have *recent* as both a space and display selector already.

**Comment by n0ibe** [comment]
Is there a way to fix this with the tools available in version 2.4.3? If recent were to become space specific I'd need to upgrade to have that functionality, but I really don't want to lose window borders.

**Comment by koekeishiya** [comment]
I don't think this is possible to do without code changes. For what it's worth I ported the yabai border functionality (with feature parity) into a standalone application: https://github.com/koekeishiya/limelight

**Comment by koekeishiya** [comment]
Actually after thinking some more, I think the *recent* selector in yabai should work the way it does, but it should work across all spaces. macOS already does track the most recent focused within a space and will automatically focus that window when you activate said space.

**Comment by n0ibe** [comment]
> For what it's worth I ported the yabai border functionality (with feature parity) into a standalone application: https://github.com/koekeishiya/limelight

Oh, didn't know that. I just built `limelight` and it works as expected, but how do I recover the option to specify `border=off` in yabai rules? 

Btw, after upgrading to version 3.0.2 I'm having problems starting yabai as a service with homebrew. Even after installing the new scripting addition, I get a `yabai-msg: failed to connect to socket error`.

I tried removing yabai from the accessibility options hoping I'd get prompted to grant access again, but that's not happening. Any advice?


> macOS already does track the most recent focused within a space and will automatically focus that window when you activate said space.

It doesn't for me, after focusing a space I have to manually focus a window in that space. That's why I added `yabai -m window --focus recent` to my space-switching bindings, I hoped it would solve that problem.

**Comment by n0ibe** [comment]
> It doesn't for me, after focusing a space I have to manually focus a window in that space.

I found out the reason for that is I have unchecked the option `When switching to an application, switch to a Space with open windows for the application` in the mission control preferences. 

That's Apple fault's though, there's a difference between not switching to a space with windows open for the focused application, and not focusing the most recent window when switching spaces.

**Comment by koekeishiya** [comment]
>  but how do I recover the option to specify border=off in yabai rules?

You'd have to port the window rules system from yabai to limelight I guess.

> Btw, after upgrading to version 3.0.2 I'm having problems starting yabai as a service with homebrew. Even after installing the new scripting addition, I get a yabai-msg: failed to connect to socket error.

Does it work if you run it directly from the terminal? Are there any crash logs (check Console.app > Crash Reports). Does brew work if you install the *--head* version?

**Comment by n0ibe** [comment]
Installing from head worked, but I think I just needed to manually add the yabai binary in the accessibility apps.

> That's Apple fault's though, there's a difference between not switching to a space with windows open for the focused application, and not focusing the most recent window when switching spaces.

Is there a way to fix this? I'd like to leave that checked, but without losing the ability to focus the most recent window when I switch to a different space.

**Comment by koekeishiya** [comment]
After the space switch has occurred, you can call yabai -m query --windows --space  and grab the id of the first entry in the list and pass it to yabai -m window --focus <id here> to focus the most recently focused window in that space.

**Comment by noib3** [comment]
I'm sorry to reopen this issue, but I'm still struggling with focusing the most recent window on a space when switching spaces. It's quite a bit annoying having to manually focus a window (with `cmd-tab` or the mouse) every single time I switch a space.

I tried the solution you suggested in your last reply, i.e.

```
alt - 1 : yabai -m space --focus 1 && yabai -m window --focus "$(yabai -m query --windows --space | jq .[0].id)"
alt - 2 : yabai -m space --focus 2 && yabai -m window --focus "$(yabai -m query --windows --space | jq .[0].id)"
```

but it doesn't seem to work. Btw, I still have `When switching to an application, switch to a Space with open windows for the application` unchecked in System Preferences -> Mission Control.

**Comment by alexwforsythe** [comment]
Note: fixed in #719
